File: ruby/04Factory/factory.rb
#!/usr/bin/ruby -Ks

class PizzaStore
  def orderPizza( type )
    @pizza
    @pizza = createPizza( type )
    @pizza.prepare
    @pizza.bake
    @pizza.cut
    @pizza.box
    @pizza
  end
  
  def createPizza( type )
  end
end

class NYPizzaStore < PizzaStore
  def createPizza( item )
    case item
    when "チーズ"
      NYStyleCheesePizza.new
##    when "野菜"
##      NYStyleVeggiePizza.new
##    when "クラム"
##      NYStyleClamPizza.new
##    when "ペパロニ"
##      NYStylePepperoniPizza.new
    else
      nil
    end
  end
end

class ChicagoPizzaStore < PizzaStore
  def createPizza( item )
    case item
    when "チーズ"
      ChicagoStyleCheesePizza.new
##    when "野菜"
##      ChicagoStyleVeggiePizza.new
##    when "クラム"
##      ChicagoStyleClamPizza.new
##    when "ペパロニ"
##      ChicagoStylePepperoniPizza.new
    else
      nil
    end
  end
end

class CaliforniaPizzaStore < PizzaStore
  def createPizza( item )
    case item
    when "チーズ"
      CaliforniaStyleCheesePizza.new
    when "野菜"
      CaliforniaStyleVeggiePizza.new
    when "クラム"
      CaliforniaStyleClamPizza.new
    when "ペパロニ"
      CaliforniaStylePepperoniPizza.new
    else
      nil
    end
  end
end

class Pizza
  @name
  @dough
  @sauce
  @toppings
  
  def prepare
    puts "#{@name}を下処理"
    puts "生地をこねる..."
    puts "ソースを追加..."
    puts "トッピングを追加:"
    @toppings.each {|t|
      puts " #{t}"
    }
  end
  
  def bake
    puts "350度で25分間焼く"
  end
  
  def cut
    puts "ピザを扇形に切り分ける"
  end
  
  def box
    puts "PizzaStore の正式な箱にピザを入れる"
  end
  
  def getName
    @name
  end
end

class NYStyleCheesePizza < Pizza
  def initialize
    @name = "ニューヨークスタイルのソース&チーズピザ"
    @dough = "薄いクラスト生地"
    @sauce = "マリナラソース"
    @toppings = []
    @toppings << "粉レッジャーノチーズ"
  end
end

class ChicagoStyleCheesePizza < Pizza
  def initialize
    @name = "シカゴスタイルのディープディッシュチーズピザ"
    @dough = "極厚クラスト生地"
    @sauce = "プラムトマトソース"
    @toppings = []
    @toppings << "刻んだモッツァレラチーズ"
  end
  
  def cut
    puts "ピザを四角形に切り分ける"
  end
end

def main
  nyStore = NYPizzaStore.new
  chicagoStore = ChicagoPizzaStore.new
  pizza = nyStore.orderPizza( "チーズ" )
  puts "イーサンの注文は#{pizza.getName}"
  pizza = chicagoStore.orderPizza( "チーズ" )
  puts "ジョエルの注文は#{pizza.getName}"
end

main

File: python/factory.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Head First デザインパターンを写経する - 4章 Factory パターン - ヨタの日々(2009-05-06)
# http://www.area51.gr.jp/~rin/diary/?date=20090506#p04

class PizzaStore:
  def orderPizza(self, _type):
    self.pizza = self.createPizza(_type)
    self.pizza.prepare()
    self.pizza.bake()
    self.pizza.cut()
    self.pizza.box()
    return self.pizza

  def createPizza(self, _type): pass

class NYPizzaStore(PizzaStore):
  def createPizza(self, item):
    if item == "チーズ":
      return NYStyleCheesePizza()
##    elif item == "野菜":
##      return NYStyleVeggiePizza()
##    elif item == "クラム":
##      return NYStyleClamPizza()
##    elif item == "ペパロニ":
##      return NYStylePepperoniPizza()
    else:
      return None

class ChicagoPizzaStore(PizzaStore):
  def createPizza(self, item):
    if item == "チーズ":
      return ChicagoStyleCheesePizza()
##    elif item == "野菜":
##      return ChicagoStyleVeggiePizza()
##    elif item == "クラム":
##      return ChicagoStyleClamPizza()
##    elif item == "ペパロニ":
##      return ChicagoStylePepperoniPizza()
    else:
      return None

class CaliforniaPizzaStore(PizzaStore):
  def createPizza(self, item):
    if item == "チーズ":
      return CaliforniaStyleCheesePizza()
##    elif item == "野菜":
##      return CaliforniaStyleVeggiePizza()
##    elif item == "クラム":
##      return CaliforniaStyleClamPizza()
##    elif item == "ペパロニ":
##      return CaliforniaStylePepperoniPizza()
    else:
      return None


class Pizza:
  def __init__(self):
    self.name = None
    self.dough = None
    self.sauce = None
    self.toppings = None

  def prepare(self):
    print "%s を下処理" % (self.name)
    print "生地をこねる..."
    print "ソースを追加..."
    print "トッピングを追加:"
    for t in self.toppings:
      print " %s" % (t)

  def bake(self):
    print "350度で25分間焼く"

  def cut(self):
    print "ピザを扇形に切り分ける"

  def box(self):
    print "PizzaStore の正式な箱にピザを入れる"

  def getName(self):
    return self.name

class NYStyleCheesePizza(Pizza):
  def __init__(self):
    self.name = "ニューヨークスタイルのソース&チーズピザ"
    self.dough = "薄いクラスト生地"
    self.sauce = "マリナラソース"
    self.toppings = []
    self.toppings.append("粉レッジャーノチーズ")

class ChicagoStyleCheesePizza(Pizza):
  def __init__(self):
    self.name = "シカゴスタイルのディープディッシュチーズピザ"
    self.dough = "極厚クラスト生地"
    self.sauce = "プラムトマトソース"
    self.toppings = []
    self.toppings.append("刻んだモッツァレラチーズ")

  def cut(self):
    print "ピザを四角形に切り分ける"

def main():
  nyStore = NYPizzaStore()
  chicagoStore = ChicagoPizzaStore()
  pizza = nyStore.orderPizza("チーズ")
  print "イーサンの注文は %s" % (pizza.getName())
  pizza = chicagoStore.orderPizza("チーズ")
  print "ジョエルの注文は %s" % (pizza.getName())

main()
